### PR TITLE
Painless: Only allow Painless type names to be the same as the equivalent Java class.

### PR DESCRIPTION
**Pull request by jdconrad**
Updated the WhitelistLoader to reflect the changes along with current whitelist resource files.

When defining a Painless struct, the line will now be ```class java.lang.String {``` meaning java.lang.String will be available as Painless type.

If we want the short name of any class the line will now be ```class java.lang.String import {``` meaning both java.lang.String and String will be available as a Painless type.

This simplifies Painless whitelists in a significant way moving forward as we can now infer a Painless type name from a Java Class object.  Also enforces a 1 to 1 Painless type to Java class naturally.

### CODE REVIEW

**Comment by rjernst** [comment]
One high level suggestion: I think most of the time we will use the simple class names. So instead of needing to add `import` for every class, why not flip it and have a special identifier for those we want to only allow with the fully qualified name? There are very few of these (only classes that are "internal" that must be whitelisted so their methods are available, but constructing them is not expected).  I would be very explicit about this and replace `import` in your syntax with `only_fqn` (only full qualified name).

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/Definition.java [hunk 1]
```diff
@@ -678,8 +704,17 @@ private Definition(List<Whitelist> whitelists) {
     }
 
     private void addStruct(ClassLoader whitelistClassLoader, Whitelist.Struct whitelistStruct) {
-        if (!whitelistStruct.painlessTypeName.matches("^[_a-zA-Z][._a-zA-Z0-9]*")) {
-            throw new IllegalArgumentException("invalid struct type name [" + whitelistStruct.painlessTypeName + "]");
+        String painlessTypeName = whitelistStruct.javaClassName.replace('$', '.');
+        String importedPainlessTypeName = painlessTypeName;
+
+        if (painlessTypeName.matches("^[_a-zA-Z][._a-zA-Z0-9]*") == false) {
```
Comment (rjernst) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/Definition.java]: Maybe move this out to a private static constant so the regex is not recompiled for every struct in the whitelist?

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/Definition.java [hunk 2]
```diff
@@ -698,22 +733,41 @@ private void addStruct(ClassLoader whitelistClassLoader, Whitelist.Struct whitel
...
+            if (existingStruct == null) {
+                structsMap.put(importedPainlessTypeName, structsMap.get(painlessTypeName));
+                Type painlessType = simpleTypesMap.remove(painlessTypeName);
+
+                if (simpleTypesMap.containsKey(importedPainlessTypeName) == false) {
+                    simpleTypesMap.put(importedPainlessTypeName, painlessType);
+                }
+            } else if (existingStruct.clazz.equals(javaClass) == false) {
+                throw new IllegalArgumentException("imported name [" + painlessTypeName + "] is used to " +
+                    "illegally represent multiple java classes [" + whitelistStruct.javaClassName + "] " +
```
Comment (rjernst) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/Definition.java]: nit: I would reword this, as "illegally" sounds odd. Perhaps something like:

`Simple type name [Foo] found for types [my.Foo] and [other.Foo]. At least one must require fully qualified name using "only_fqn"`

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/WhitelistLoader.java [hunk 3]
```diff
@@ -43,18 +43,29 @@
      * and field.  Most validation will be done at a later point after all white-lists have been gathered and their
      * merging takes place.
      *
+     * A painless type name is considered to one of the following:
```
Comment (rjernst) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/WhitelistLoader.java]: `is considered to one of` -> `is one of`

Comment (rjernst) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/WhitelistLoader.java]: Also, maybe make it clear this is explained so that these types can be used in method signatures. The class portion of the whitelist syntax must contain the fully qualified name.

**Review by rjernst — COMMENTED** [review]
I left some minor suggestions.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/WhitelistLoader.java (hunk 3 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/WhitelistLoader.java]: Done.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/Definition.java (hunk 2 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/Definition.java]: Changed.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/Definition.java (hunk 1 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/Definition.java]: Done.

**Comment by jdconrad** [comment]
@rjernst Changed import to only_fqn and responded to the rest of the PR comments.  Would you please take another look?  Thanks!

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/Definition.java [hunk 4]
```diff
@@ -33,13 +33,16 @@
...
 import java.util.Stack;
+import java.util.regex.Pattern;
 
 /**
  * The entire API for Painless.  Also used as a whitelist for checking for legal
  * methods and fields during at both compile-time and runtime.
  */
 public final class Definition {
 
+    private static final Pattern TYPE_NAME_PATTERN = Pattern.compile("^[_a-zA-Z][._a-zA-Z0-9]*$");
```
Comment (rjernst) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/Definition.java]: A type name can start with an underscore?

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/Definition.java [hunk 5]
```diff
@@ -577,7 +583,13 @@ private Definition(List<Whitelist> whitelists) {
 
         // goes through each Painless struct and determines the inheritance list,
         // and then adds all inherited types to the Painless struct's whitelist
-        for (Struct painlessStruct : structsMap.values()) {
+        for (String painlessStructName : structsMap.keySet()) {
```
Comment (rjernst) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/Definition.java]: What's the reason for moving to key iteration followed by lookup and filtering? Iterating values + filtering should be faster (not that performance matters here really, but...).

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/Definition.java [hunk 6]
```diff
@@ -698,22 +736,41 @@ private void addStruct(ClassLoader whitelistClassLoader, Whitelist.Struct whitel
...
                     "illegally represent multiple java classes [" + whitelistStruct.javaClassName + "] and " +
                     "[" + existingStruct.clazz.getName() + "]");
         }
+
+        if (whitelistStruct.onlyFQNJavaClassName == false) {
+            existingStruct = structsMap.get(importedPainlessTypeName);
+
+            if (existingStruct == null) {
+                structsMap.put(importedPainlessTypeName, structsMap.get(painlessTypeName));
+                Type painlessType = simpleTypesMap.remove(painlessTypeName);
```
Comment (rjernst) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/Definition.java]: I'm a little confused by this block. Why is the type removed only to be added back?

**Review by rjernst — COMMENTED** [review]
This looks good I just have one last outstanding question and a couple other minor comments.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/Definition.java (hunk 4 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/Definition.java]: Yes.  This matches the ANTLR grammar.  This can probably be changed, but I don't want to add more complications to grammar ID's right now.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/Definition.java (hunk 5 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/Definition.java]: Was for more explicit naming, but changed to use entryset to get the values from.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/Definition.java (hunk 6 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/Definition.java]: For the doc generator to work simple types names must not contain multiple entries of the same type with multiple names, so this removes the fully-qualified name in favor the short name for now to be consistent with what already exists.  This definitely needs to be cleaned up but not in this PR.

**Comment by jdconrad** [comment]
@rjernst Responded to all PR comments.

**Comment by jdconrad** [comment]
@rjernst Fixed the inconsistency we chatted about.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/Definition.java [hunk 7]
```diff
@@ -747,30 +747,24 @@ private void addStruct(ClassLoader whitelistClassLoader, Whitelist.Struct whitel
...
-            } else if (existingStruct.clazz.equals(javaClass) == false) {
-                throw new IllegalArgumentException("simple type name [" + painlessTypeName + "] found for types " +
-                    "[" + whitelistStruct.javaClassName + "] and [" + existingStruct.clazz.getName() + "]; " +
-                    "at least one must require fully qualified name using 'only_fqn'");
-            }
+        } else if (whitelistStruct.onlyFQNJavaClassName && simpleTypesMap.containsKey(importedPainlessTypeName) &&
+                simpleTypesMap.get(importedPainlessTypeName).clazz == javaClass ||
+                whitelistStruct.onlyFQNJavaClassName == false && (simpleTypesMap.containsKey(importedPainlessTypeName) == false ||
+                simpleTypesMap.get(importedPainlessTypeName).clazz != javaClass)) {
+            throw new IllegalArgumentException("inconsistent only_fqn parameters specified for type [" + painlessTypeName + "]");
```
Comment (rjernst) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/Definition.java]: This message might be confusing since this could happen across whitelists in different plugins right? Maybe expand it a little, or at least say "found" instead of "specified"?

**Review by rjernst — APPROVED** [review]
LGTM, just one last suggestion

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/Definition.java (hunk 7 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/Definition.java]: Changed.

**Comment by jdconrad** [comment]
@elasticmachine Please test this.

**Comment by jdconrad** [comment]
@elasticmachine test this please

**Comment by jdconrad** [comment]
@rjernst Thanks for the review!
